### PR TITLE
Add logging context to `PreConfigureCallback` and `PreCheckCallback`

### PR DESCRIPTION
**Pull request by iwahbe**
Add the ability to return structured logs from `Pre*Callback` methods.

---

Side note:

We should deprecate `PreConfigureCallbackWithLogger` for `PreConfigureCallbackWithCtx` with an attached logger. This will fix the interface (not directly exposing callers to `*provider.Host`) and allow us to attach additional context data in a consistent way.

---

This PR contains **BREAKING CHANGES** to the following functions. All three functions now take a `context.Context` as their first argument. None of the functions are used outside of the bridge, according to a GH code search:
- `tfbridge.MakeTerraformInputs`: https://github.com/search?q=%2FMakeTerraformInputs%2F+-is%3Afork&type=code
- `tfbridge.MakeTerraformState`: https://github.com/search?q=%2FMakeTerraformState%2F+-is%3Afork&type=code
- `tfbridge.MakeTerraformConfig`: https://github.com/search?q=%2FMakeTerraformConfig%2F+-is%3Afork&type=code

Given the unique context where these functions make sense, I believe they are unlikely to be used outside the bridge.

### CODE REVIEW

File: pkg/tfbridge/info.go
Comment (t0yv0) [inline pkg/tfbridge/info.go:128]: Could we use the existing interface from the Go SDK here?
 https://github.com/pulumi/pulumi/blob/ee25d63598eae99b4bc170d7b0595fd8609abf50/sdk/go/pulumi/log.go#L29-#L29 

What is Status()? Is it the same as LogArgs.Ephemeral = true?

File: pkg/tfbridge/log.go
Comment (t0yv0) [inline pkg/tfbridge/log.go:131]: The err accumulator is eyebrow-raising. Why would you do that? What kind of errors are you trying to accumulate here?

File: pkg/tfbridge/log.go
Comment (t0yv0) [inline pkg/tfbridge/log.go:141]: Looks like we're accumulating errors from being unable to send a log message to the engine, perhaps because the engine is shutting down. Why? Would we ever want to change control flow behavior based on not being able to reach the engine? An alternative here is to dump it to glog low-level stream at verbose 9 level and proceed as if nothing happened.

File: pkg/tfbridge/provider.go
Comment (t0yv0) [inline pkg/tfbridge/provider.go:333]: I'm not sure I understand error handling here, are we trying to promote accumulated errors to an error result from Check? What's an example where we need to do this?

**Comment by t0yv0** [comment]
This is great directionally, love utilizing Context more for propagating the logger interface. Had some questions above.

File: pkg/tfbridge/info.go
Comment (iwahbe) [reply pkg/tfbridge/info.go:128]: We can't use that interface because it expects `Resource`s, not `URN`s. It's at the wrong level.

`Status()` is the same `LogArgs.Ephemeral = true`. The normal `provider.Host.Log` has `Debug`, `DebugStatus`, etc arguments. That's where I took the name from.

File: pkg/tfbridge/log.go
Comment (t0yv0) [reply pkg/tfbridge/log.go:131]: Why ever collect them if we don't want to expose them? What do you want to do with them? Can they just be flushed to glog and forgotten?

File: pf/tfbridge/provider_checkconfig.go
Comment (iwahbe) [reply pf/tfbridge/provider_checkconfig.go:78]: We do `PreCheckCallback` for sdk and pf. We do `PreConfigureCallback` for sdk and pf. I'm adding `ComputeDefault`. Can you think of any other places?

File: pkg/tfbridge/info.go
Comment (t0yv0) [reply pkg/tfbridge/info.go:128]: Ah yes. Well there's something to be said about making codebases more independent by not borrowing interfaces anyway. Thanks!

**Comment by t0yv0** [comment]
The breaking changes are no-go I think. We have large customer codebases that use the bridge heavily that are not part of public GH code search. If we need to do this we can add tfbridge.MakeTerraformInputsWithContext variants and move prod to use them, deprecating previous ones.

File: pf/tfbridge/provider_checkconfig.go
Comment (iwahbe) [reply pf/tfbridge/provider_checkconfig.go]: I have injected the logger at the root context of all provider gRPC calls. This means that it should be everywhere (that is correctly threaded from the original context).

**Comment by t0yv0** [comment]
```
error: resource or variable named "unfairoRandomNumber" could not be found

  on Pulumi.yaml line 61:
  61:   unfairRandomNumberFair__actual: ${unfairoRandomNumber.fair}

```

I think this depends on https://github.com/pulumi/pulumi-terraform-bridge/pull/1336 then. Let's see.

File: pkg/tfbridge/info.go
Comment (t0yv0) [inline pkg/tfbridge/info.go:131]: Nit: if there's any pointers on what these are or how they're different, perhaps a link, that'd be great; but not blocking on this.

**Review by t0yv0 — APPROVED** [review]
LGTM once we get the tests to pass. Double-checked the dependent code corpus analysis and it seems like we're OK taking a minor break on the functions that de-facto are used internally - the MakeTerraformInput and so on. We will need to call it out in real notes and revert if there's any complaints.

File: pkg/tfbridge/info.go
Comment (iwahbe) [reply pkg/tfbridge/info.go:131]: Opened https://github.com/pulumi/pulumi-terraform-bridge/issues/1338 to track. It's late in the day and I want to get an alpha out today, so I'll come back to this later.

**Pull request merged**
